File: src/sprite.c
#include <stdio.h>
#include "sprite.h"

const unsigned int TEX_WIDTH = 512;
const unsigned int TEX_HEIGHT = 512;

tex_quad get_current_frame(animation *anim) {
	return anim->frames[(i32)anim->current_frame];
}

quad get_quad_from_sprite(sprite *spr) {
	printf("Get quad from sprite\n");
	vec3 pos = spr->pos;
	tex_quad frame;
	if (spr->animated) {
		frame = get_current_frame(&spr->anim);
	} else {
		frame = spr->tex;
	}

	return new_quad(
		new_vertex(
			pos,
			frame.tl
		),
		new_vertex(
			new_vec3(pos.x + spr->width, pos.y, pos.z),
			frame.tr
		),
		new_vertex(
			new_vec3(pos.x, pos.y + spr->height, pos.z),
			frame.bl
		),
		new_vertex(
			new_vec3(pos.x + spr->width, pos.y + spr->height, pos.z),
			frame.br
		)
	);
}

static void animate(animation *anim, f64 delta) {
	anim->current_frame += (anim->framerate * delta);
	if (anim->current_frame >= anim->frame_count) {
		// TODO (etate): Potential edge case. If enough time goes by
		// between renders, we could still exceed the frame_count after
		// this operation.
		anim->current_frame = anim->current_frame - anim->frame_count;
	}
}

void sprite_animate(sprite *spr, f64 delta) {
	// short circuit if not animated
	if (!spr->animated) {
		return;
	}

	animation *anim = &spr->anim;
	i32 previous_frame = (i32)anim->current_frame;
	animate(anim, delta);

	// short circuit if the frame hasn't changed
	if ((i32)anim->current_frame == previous_frame) {
		return;
	}

	// update sprite quad with new texture coordinates
	tex_quad new_tex = anim->frames[(i32)anim->current_frame];
	spr->_quad->tl.tex_coord = new_tex.tl;
	spr->_quad->tr.tex_coord = new_tex.tr;
	spr->_quad->bl.tex_coord = new_tex.bl;
	spr->_quad->br.tex_coord = new_tex.br;
}

sprite new_sprite(vec3 pos, int width, int height, bool animated) {
	return (sprite){
		.id = 0,
		.pos = pos,
		.width = width,
		.height = height,
		.animated = animated,
		._quad = NULL,
	};
}

void set_sprite_pos(sprite *spr, vec3 pos) {
	spr->pos = pos;
	quad *q = spr->_quad;

	q->tl.pos = pos;
	q->tr.pos = new_vec3(pos.x + spr->width, pos.y, pos.z);
	q->bl.pos = new_vec3(pos.x, pos.y + spr->height, pos.z);
	q->br.pos = new_vec3(pos.x + spr->width, pos.y + spr->height, pos.z);
}

vec2 new_tex_coord(f32 x, f32 y, i32 width, i32 height) {
	return vec2_to_texture_space(new_vec2(x, y), width, height);
}

animation new_animation(f32 framerate, i32 frame_count, vec2 offset, i32 frame_width, i32 frame_height, i32 tex_width, i32 tex_height) {
	tex_quad *frames = (tex_quad *)malloc(sizeof(tex_quad) * 30);
	for (int i = 0; i < frame_count; i++) {
		frames[i] = (tex_quad){
			.tl = new_tex_coord(offset.x + (i * frame_width), offset.y, tex_width, tex_height),
			.tr = new_tex_coord(offset.x + (i * frame_width) + frame_width, offset.y, tex_width, tex_height),
			.bl = new_tex_coord(offset.x + (i * frame_width), offset.y + frame_height, tex_width, tex_height),
			.br = new_tex_coord(offset.x + (i * frame_width) + frame_width, offset.y + frame_height, tex_width, tex_height),
		};
	}

	return (animation){
		.framerate = framerate,
		.frame_count = frame_count,
		.frames = frames,
	};
}

File: src/sprite.rs
use crate::gm::{Quad, Rect, Vec2, Vec3, Vertex};
use crate::texture;

#[derive(Clone, Debug)]
pub struct Animation {
    framerate: f32,
    current_frame: f32,
    frames: Vec<texture::Quad>,
}

#[derive(Clone, Debug)]
pub enum Show {
    Tex(texture::Quad),
    Anim(Animation),
}

#[derive(Clone, Debug)]
pub struct Sprite {
    pub id: u32,
    pub pos: Vec3,
    pub width: u32,
    pub height: u32,
    pub show: Show,
    pub solid: bool,
    pub flip: bool,
    hb: Rect,
}

impl Sprite {
    pub fn new(
        id: u32,
        pos: Vec3,
        width: u32,
        height: u32,
        hb: Rect,
        solid: bool,
        show: Show,
    ) -> Sprite {
        Sprite {
            id,
            pos,
            width,
            height,
            hb,
            solid,
            flip: false,
            show,
        }
    }

    pub fn to_quad(&self) -> Quad {
        let mut tex_quad = self.tex_quad();
        if self.flip {
            let tl = tex_quad.tl;
            let bl = tex_quad.bl;
            tex_quad.tl = tex_quad.tr;
            tex_quad.tr = tl;
            tex_quad.bl = tex_quad.br;
            tex_quad.br = bl;
        }

        Quad::new(
            Vertex::new(self.pos, tex_quad.tl),
            Vertex::new(
                Vec3::new(self.pos.x + self.width as f32, self.pos.y, self.pos.z),
                tex_quad.tr,
            ),
            Vertex::new(
                Vec3::new(self.pos.x, self.pos.y + self.height as f32, self.pos.z),
                tex_quad.bl,
            ),
            Vertex::new(
                Vec3::new(
                    self.pos.x + self.width as f32,
                    self.pos.y + self.height as f32,
                    self.pos.z,
                ),
                tex_quad.br,
            ),
        )
    }

    pub fn tex_quad(&self) -> texture::Quad {
        match &self.show {
            Show::Tex(tex) => tex.clone(),
            Show::Anim(anim) => anim.tex_quad(),
        }
    }

    pub fn set_pos(&mut self, pos: Vec3) {
        self.pos = pos;
    }

    pub fn tick(&mut self, delta: f32) {
        match &mut self.show {
            Show::Anim(anim) => anim.tick(delta),
            _ => (),
        }
    }

    pub fn hitbox(&self) -> Rect {
        Rect::new(self.pos + self.hb.pos, self.hb.w, self.hb.h)
    }

    pub fn will_overlap(&self, move_vec: Vec3, hb: &Rect) -> bool {
        let future_hb = Rect::new(self.pos + move_vec + self.hb.pos, self.hb.w, self.hb.h);

        future_hb.overlaps(hb)
    }

    pub fn show(&mut self, show: Show) {
        self.show = show
    }
}

impl Animation {
    pub fn new(atlas: &texture::Atlas, framerate: f32, frames: Vec<usize>) -> Animation {
        Animation {
            framerate,
            current_frame: 0.0,
            frames: frames
                .iter()
                .copied()
                .map(|frame| atlas.get_frame_by_idx(frame))
                .collect(),
        }
    }

    pub fn tex_quad(&self) -> texture::Quad {
        self.frames[self.current_frame as usize]
    }

    pub fn tick(&mut self, delta: f32) {
        self.current_frame += self.framerate * (delta as f32);
        if self.current_frame > self.frames.len() as f32 {
            self.current_frame -= self.frames.len() as f32;
        }
    }
}
